### PR TITLE
Shadow ban

### PR DESCRIPTION
**Pull request by arpit270896**

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -1,5 +1,6 @@
 SESSION_SECRET=
 OUTGOING_TOKEN=
+OUTGOING_TOKEN_BOT=
```
Comment (apsdehal) [inline .env.sample]: More descriptive name please! Something like `ADMIN_TOKEN`.

File: user.coffee [hunk 2]
```diff
@@ -0,0 +1,61 @@
+fs = require('fs')
+
+reservedNicks = fs.readFileSync('reserved_nicks').toString().trim().split('\n')
+nick_session_map = {}
+banned_sessions = []
+status = {}
+
+# Verify the nick of user
+verify_nick = (nick)->
```
Comment (apsdehal) [inline user.coffee]: Consistent styles please, jinora (well 90% of JS) uses camelCase.

File: user.coffee [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  for reservedNick in reservedNicks
+    if nick.toLowerCase().indexOf(reservedNick) != -1
+      return false
+
+  return true
+
+# Verify the session of user
+verify_session = (session_id)->
+  if session_id == null or session_id == undefined or session_id == ""
```
Comment (apsdehal) [inline user.coffee]: change this to `typeof sessionId == undefined or sessionId.length == 0`

File: user.coffee [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  if session_id in banned_sessions
+    return false
+
+  return true
+
+
+module.exports = {
+  # Ban nick when nick is posted to /user/nick/ban
+  ban_nick: (nick)->
+    if nick == null or nick == undefined or nick == ""
```
Comment (apsdehal) [inline user.coffee]: Same for here. Instead convert this to function.

File: user.coffee [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    fs.appendFile 'reserved_nicks', nick + "\n", (err)->
+      return false if err
+    reservedNicks.push nick
+
+    return true
+
+  # Ban session when nick is posted to /user/session/ban
+  ban_session: (nick)->
+    if nick == null or nick == undefined or nick == ""
```
Comment (apsdehal) [inline user.coffee]: Here too!

File: user.coffee [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      return false
+
+    if nick_session_map[nick] != undefined
+      banned_sessions.push nick_session_map[nick] if banned_sessions[nick_session_map[nick]] == undefined
```
Comment (apsdehal) [inline user.coffee]: This whole check makes now sense. See it like this:

bannedSessions = [];
bannedSession.push('a'); 
// bannedSession = ['a'];
// So checking bannedSession['a'] would always returned undefined. You are mixing up objects with arrays.

File: public/polymer.coffee [hunk 7]
```diff
@@ -64,6 +62,9 @@ socket.on 'connect', ->
   socket.emit 'presence:demand'
 
 socket.on 'chat:msg', (msg)->
+  if msg.status['session'] and !msg.status['nick']
+    msg.nick = template.userName = prompt 'Sorry! You can\'t have this username. \nPlease enter another username'
```
Comment (apsdehal) [inline public/polymer.coffee]: Can you find a better way to do this. This is close to a hack.

File: user.coffee (hunk 3 above)
Comment (captn3m0) [inline user.coffee]: `typeof sessionId === 'undefined'`, no?

Comment (apsdehal) [inline user.coffee]: Yes :)

File: app.coffee [hunk 8]
```diff
@@ -54,14 +55,40 @@ app.post "/webhook", (req, res) ->
...
+
+# Ban the session
+app.post "/user/session/ban", (req, res)->
+  throw "Invalid Token" unless req.body.token == process.env.OUTGOING_TOKEN_BOT
+
+  if user.ban_session req.body.nick
+    res.send "ok"
+    return
+
+  res.send "error!"
```
Comment (apsdehal) [inline app.coffee]: Use if else instead of premature returning.

**Comment by apsdehal** [comment]
Final thing, update README to explain ban features

File: user.coffee (hunk 3 above)
Comment (arpit270896) [inline user.coffee]: CoffeeScript compiles `==` into `===`, and `!=` into  `!==`.

Comment (arpit270896) [inline user.coffee]: and `typeof sessionId == undefined` or `sessionId.length == 0` gives error when sessionId is set to null

Comment (apsdehal) [inline user.coffee]: Nemo meant `'undefined'` instead of `undefined`

Comment (captn3m0) [inline user.coffee]: I forgot that you were doing coffeescript. This doesn't feel like coffeescript at all. See http://stackoverflow.com/a/10238211/368328

And probably read through the coffeescript cookbook as well.

Comment (apsdehal) [inline user.coffee]: And best solution to this is
```js
if sessionId
 // do something 
```
This will work in all cases.

File: app.coffee (hunk 8 above)
Comment (captn3m0) [inline app.coffee]: What is this? All slack messages (that reach jinora via slack) are already authenticated, iirc. There is no need to create a new endpoint for ban. Just have a keyword match for slack messaages instead:

`jinora ban admin`

And the user would be banned.

Comment (arpit270896) [inline app.coffee]: I was thinking of integrating bannig with bot.
Something like `bot ban admin`

Comment (captn3m0) [inline app.coffee]: jinora already receives all messages that we type in the public channel. Why add needless complexity by bringing in another dependency that not every might have installed.

Comment (arpit270896) [inline app.coffee]: At first I thought the same, but if there's a small typo while typing `jinora ban admin`, the message would reach the users, which destroys the purpose of shadow banning.

Comment (captn3m0) [inline app.coffee]: Then make it a slash command in slack.

**Comment by captn3m0** [comment]
@arpit270896 I thought of a way to make sure of the ban feature. Once a user types `jinora ban [X]`, jinora should ask for confirmation from that very same user. The next message should be a yes. If its anything else, it ignores the ban.

**Comment by arpit270896** [comment]
Do you mean any jinora user can ban another user ?

**Comment by captn3m0** [comment]
No, every slack user can ban any anon user.
It goes like this:

```cat -> spams
nemo -> jinora ban cat (S)
jinora -> nemo, are you sure you wan't to ban cat? (S)
nemo -> yes (S)
jinora -> nemo just banned cat (S)
cat -> spams, but is posted to public-banned```

Messages marked with (S) are only visible in slack.

**Comment by arpit270896** [comment]
@captn3m0 any particular reason to ask for confirmation, can't we just do it like this:
`cat -> spams
nemo -> !ban user cat (S)
jinora -> nemo just banned cat (S)
cat -> spams but posted to public channel
nemo -> !ban usr cat (S)                          (typo here)
jinora -> Invalid command (S)
nemo -> !unban user cat (S)
jinora -> nemo just unbanned cat (S)
`

**Comment by captn3m0** [comment]
Yeah, I think you can mark bang as a special seeing as we already do private messages via that.

**Comment by captn3m0** [comment]
Also,just thought bans won't stay forever, so if server restarts, it will get reset.

**Comment by arpit270896** [comment]
Also I have currently implemented, session ban. So even on browser restart, the bans would go.
But a user would try that only if he comes to know about that ban, which he will not(in most of the cases).

**Comment by captn3m0** [comment]
Not, really. If you keep encrypted sessions  in the client, it should be doable.

**Comment by apsdehal** [comment]
Closed in 7b212bc

**Pull request closed without merging**